popde/adaptive_kde.py

```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        self.alpha_options = alpha_options
+        self.bandwidth_options = bandwidth_options
+        super().__init__(data, weights, input_transf, stdize,
+                         rescale, backend, bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, bandwidth_val, alpha_val):
+        loo = LeaveOneOut() #sklearn way
+        weights = self.weights
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+           
+            #######There may need some hints how to use it better 
+            ### also if we need to add weight option here
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom += awkde.evaluate(test_data)  
+        return np.mean(fom)
+
+    def kfold_cv_score(self, bandwidth_val, alpha_val, n_splits=2):
+        """
+        to do: how to choose and use s_split in init
+        """
+        kf = KFold(n_splits=n_splits, shuffle=True, random_state=42)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom.append(awkde.evaluate(test_data).sum())
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot=False):
+        best_score = float('inf')
+        best_params = {'bandwidth': None, 'alpha': None}
+
+        FOM= {}
+        for bandwidth in self.bandwidth_options:
+            for alpha in self.alpha_options:
+                if method=='kfold_cv':
+                    score = self.kfold_cv_score(bandwidth, alpha, n_splits=2)
+                else:
+                    score = self.loo_cv_score(bandwidth, alpha)
+
+                FOM[(bandwidth, alpha)] = score
+                if score < best_score:
+                    best_score = score
+                    best_params['bandwidth'] = bandwidth
+                    best_params['alpha'] = alpha
+        
+        if fom_plot==True:
+            import operator
+            optval = max(FOM.items(), key=operator.itemgetter(1))[0]
+            optbw, optalpha  = optval[0], optval[1]
+            maxFOM = FOM[(optbw, optalpha)]
+
+            fig = plt.figure(figsize=(12,8))
+            ax = fig.add_subplot(111)
+            for bw in self.bandwidth_options:
+                FOMlist = [FOM[(bw, al)] for al in self.alpha_options]
+                if bw not in ['silverman', 'scott']:
+                    bw = float(bw) #bwchoice.astype(np.float) #for list
+                    ax.plot(self.alpha_options, FOMlist, label='{0:.3f}'.format(bw))
+                else:
+                    ax.plot(alphagrid, FOMlist, label='{}'.format(bw))
+            if optbw not in ['silverman', 'scott']:
```

Comment:
I don't think you need two separate 'if optbw' blocks (here and line 188), can you combine them?

Reply:
done. KDEpy cannot take scott and silvermann option for multidimensional KDE. need some suggestion here

Reply:
I am not sure what you need here. We do not need to support 'scott' and 'silverman' in general so I don't know what the problem is

Reply:
done